source/MaterialXTest/Runtime.cpp

```diff
@@ -1675,4 +1675,74 @@ TEST_CASE("Runtime: libraries", "[runtime]")
     REQUIRE(api->getImplementationSearchPath().find("stdlib_genglsl_unit_impl.mtlx").exists());    
 }
 
+TEST_CASE("Runtime: units", "[runtime]")
+{
+    mx::RtScopedApiHandle api;
+
+    // Load in all libraries required for materials
+    mx::FileSearchPath searchPath(mx::FilePath::getCurrentPath() / mx::FilePath("libraries"));
+    api->setSearchPath(searchPath);
+    api->loadLibrary(STDLIB);
+    // Load in stdlib twice on purpose to ensure no exception is thrown when trying to add a duplicate unit 
+    // definition 
+    api->loadLibrary(STDLIB);
+    api->loadLibrary(PBRLIB);
+    api->loadLibrary(BXDFLIB);
+
+    // Read in test document with units
+    mx::FileSearchPath testSearchPath(mx::FilePath::getCurrentPath() /
+        "resources" /
+        "Materials" /
+        "TestSuite" /
+        "stdlib" /
+        "units");
+    mx::StringVec tests{ "distance_units.mtlx",
+                         "image_unit.mtlx",
+                         "standard_surface_unit.mtlx",
+                         "texture_units.mtlx",
+                         "tiledimage_unit.mtlx" };
+    mx::RtReadOptions options;
+    options.desiredMinorVersion = 38;
+    for (auto test : tests)
+    {
+        mx::RtStagePtr stage = api->createStage(mx::RtToken("stage: " + test));
+        mx::RtFileIo fileIo(stage);
+
+        // Test read will take into account units read in via library load
+        fileIo.read(test, testSearchPath, &options);
+
+        // Test that read and write of files with units works.
+        std::stringstream inStream;
+        mx::DocumentPtr inDoc = mx::createDocument();
+        mx::XmlReadOptions readOptions;
+        readOptions.desiredMinorVersion = 38;
+        mx::readFromXmlFile(inDoc, test, testSearchPath, &readOptions);
+
+        mx::DocumentPtr outDoc = mx::createDocument();
+        std::stringstream outStream;
+        fileIo.write(outStream);
+        mx::readFromXmlStream(outDoc, outStream);
+
+        for (mx::ElementPtr elem : inDoc->traverseTree())
```

Comment:
Can't do a document compare for now since colorspace on nodes does not load/save properly. Also string comparison won't work as attributes may change order. Do a straight ValueElement to ValueElement compare between document loaded directly and document  create from runtime.